src/main/java/org/candlepin/testext/manifestgen/EntityMapper.java

```diff
@@ -0,0 +1,353 @@
+
+            if (fetched == null) {
+                fetched = new Product()
+                    .setId(pid);
+            }
+
+            return fetched;
+        });
+
+        this.mergeNonNullField(pdto.getName(), entity::setName);
+        this.mergeNonNullField(pdto.getMultiplier(), entity::setMultiplier);
+        this.mergeNonNullField(pdto.getDependentProductIds(), entity::setDependentProductIds);
+
+        // Attributes need to be converted because they have the worst encoding
+        this.mergeNonNullField(this.convertAttributes(pdto.getAttributes()), entity::setAttributes);
+
+        // This field violates conventions. Always update it.
+        entity.setDerivedProduct(this.mergeProductData(pdto.getDerivedProduct()));
+
+        // Provided products
+        Collection<ProductDTO> providedProductDtos = pdto.getProvidedProducts();
+        if (providedProductDtos != null) {
+            List<Product> providedProducts = providedProductDtos.stream()
+                .filter(Objects::nonNull)
+                .map(this::mergeProductData)
+                .toList();
+
+            entity.setProvidedProducts(providedProducts);
+        }
+
+        // product content
+        Collection<ProductContentDTO> productContentDto = pdto.getProductContent();
+        if (productContentDto != null) {
+            List<ProductContent> productContent = productContentDto.stream()
+                .filter(Objects::nonNull)
+                .map(this::convertProductContent)
+                .toList();
+
+            entity.setProductContent(productContent);
+        }
+
+        // branding
+        Collection<BrandingDTO> brandingDtos = pdto.getBranding();
+        if (brandingDtos != null) {
+            List<Branding> branding = brandingDtos.stream()
+                .filter(Objects::nonNull)
+                .map(this::convertBranding)
+                .toList();
+
+            entity.setBranding(branding);
+        }
+
+        return entity;
+    }
+
+    public EntityMapper addSubscription(SubscriptionDTO sdto) {
+        if (sdto == null) {
+            return this;
+        }
+
+        // Impl note:
+        // The org will be net-new for every request, so every pool will be created as new from the
+        // subscription data. As such, we need not worry (as much) about merge behavior. At least
+        // not on the pool proper.
+
+        Pool pool = new Pool()
+            .setOwner(this.owner)
+            .setProduct(this.mergeProductData(sdto.getProduct()))
+            .setQuantity(sdto.getQuantity())
+            .setStartDate(Util.toDate(sdto.getStartDate()))
+            .setEndDate(Util.toDate(sdto.getEndDate()))
+            .setContractNumber(sdto.getContractNumber())
+            .setAccountNumber(sdto.getAccountNumber())
+            .setOrderNumber(sdto.getOrderNumber());
+
+            // Apparently subscription DTOs don't have attributes!?
+            // .setAttributes(this.convertAttributes(sdto.getAttributes()))
```

Comment:
Just want to double check and see if you intentionally wanted to leave this commented line

Reply:
Yeah, I'm fine leaving that there for now. I feel like it's just a matter of time before someone needs to set specific attributes on the manifest pool and then we'll need to update the API accordingly to enable that line.

Reply:
Ugh, the SubscriptionDTO is defined in the openapi spec file because it is actually used in production APIs. Adding fields strictly required for testing will pollute our API. Can we add to this comment, that if anyone wants to add attributes, please create a new SubscriptionDTO only for the testext layer, or at least extend the existing SubscriptionDTO, instead of adding to it directly?

Reply:
Done.